flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/BatchJobRecoveryTest.java

```diff
@@ -0,0 +1,1154 @@
+                    transitionExecutionsState(scheduler, ExecutionState.FINISHED, SOURCE_ID);
+                });
+        runInMainThread(
+                () -> { // trigger all middle finished.
+                    transitionExecutionsState(scheduler, ExecutionState.FINISHED, MIDDLE_ID);
+                });
+        runInMainThread(
+                () -> {
+                    // trigger all sink finished.
+                    transitionExecutionsState(scheduler, ExecutionState.FINISHED, SINK_ID);
+                });
+
+        List<ExecutionAttemptID> sourceExecutions =
+                getCurrentAttemptIds(scheduler.getExecutionJobVertex(SOURCE_ID));
+        List<ExecutionAttemptID> middleExecutions =
+                getCurrentAttemptIds(scheduler.getExecutionJobVertex(MIDDLE_ID));
+        List<ExecutionAttemptID> sinkExecutions =
+                getCurrentAttemptIds(scheduler.getExecutionJobVertex(SINK_ID));
+
+        waitWriteJobFinishedEventCompleted(12);
+        runInMainThread(() -> jobEventStore.stop(false));
+
+        // start a new scheduler and try to recover.
+        AdaptiveBatchScheduler newScheduler = createScheduler(jobGraphSupplier.get());
+        startSchedulingAndWaitRecoverFinish(newScheduler);
+
+        // check source vertices' state were recovered.
+        for (ExecutionVertex vertex :
+                getExecutionVertices(SOURCE_ID, newScheduler.getExecutionGraph())) {
+            assertThat(sourceExecutions)
+                    .contains(vertex.getCurrentExecutionAttempt().getAttemptId());
+            assertThat(vertex.getExecutionState()).isEqualTo(ExecutionState.FINISHED);
+        }
+        // check middle vertices' state were recovered.
+        for (ExecutionVertex vertex :
+                getExecutionVertices(MIDDLE_ID, newScheduler.getExecutionGraph())) {
+            assertThat(middleExecutions)
+                    .contains(vertex.getCurrentExecutionAttempt().getAttemptId());
+            assertThat(vertex.getExecutionState()).isEqualTo(ExecutionState.FINISHED);
+        }
+
+        // check sink's parallelism was recovered.
+        assertThat(newScheduler.getExecutionJobVertex(SINK_ID).getParallelism())
+                .isEqualTo(DECIDED_SINK_PARALLELISM);
+        // check sink vertices' state were recovered.
+        for (ExecutionVertex vertex :
+                getExecutionVertices(SINK_ID, newScheduler.getExecutionGraph())) {
+            assertThat(sinkExecutions).contains(vertex.getCurrentExecutionAttempt().getAttemptId());
+            assertThat(vertex.getExecutionState()).isEqualTo(ExecutionState.FINISHED);
+        }
+    }
+
+    // This case will use job graph with the following topology:
+    // Source (p=5) -- POINTWISE --> Middle (p=5) -- ALLTOALL --> Sink (p=2)
+    //
+    // This test case verifies that sourceCoordinator's split assignments are not overwritten after
+    // a JM failover, despite multiple occurrences of 'partition not found' exceptions, to prevent
```

Comment:
are not overwritten -> are restored

despite multiple occurrences of 'partition not found' exceptions -> **unless**  sources are restarted(tiggered by 'partition not found' exceptions)